File: src/main/java/org/samoxive/safetyjim/discord/commands/Ban.java
package org.samoxive.safetyjim.discord.commands;

import net.dv8tion.jda.core.EmbedBuilder;
import net.dv8tion.jda.core.JDA;
import net.dv8tion.jda.core.Permission;
import net.dv8tion.jda.core.entities.*;
import net.dv8tion.jda.core.events.message.guild.GuildMessageReceivedEvent;
import net.dv8tion.jda.core.managers.GuildController;
import org.jooq.DSLContext;
import org.samoxive.jooq.generated.Tables;
import org.samoxive.jooq.generated.tables.records.BanlistRecord;
import org.samoxive.safetyjim.discord.Command;
import org.samoxive.safetyjim.discord.DiscordBot;
import org.samoxive.safetyjim.discord.DiscordUtils;
import org.samoxive.safetyjim.discord.TextUtils;
import org.samoxive.safetyjim.helpers.Pair;

import java.awt.*;
import java.util.Date;
import java.util.Scanner;

public class Ban extends Command {
    private String[] usages = { "ban @user [reason] | [time] - bans the user with specific arguments. Both arguments can be omitted" };

    @Override
    public String[] getUsages() {
        return usages;
    }

    @Override
    public boolean run(DiscordBot bot, GuildMessageReceivedEvent event, String args) {
        Scanner messageIterator = new Scanner(args);
        JDA shard = event.getJDA();

        Member member = event.getMember();
        User user = event.getAuthor();
        Message message = event.getMessage();
        TextChannel channel = event.getChannel();
        Guild guild = event.getGuild();
        Member selfMember = guild.getSelfMember();

        if (!member.hasPermission(Permission.BAN_MEMBERS)) {
            DiscordUtils.failMessage(bot, message, "You don't have enough permissions to execute this command! Required permission: Ban Members");
            return false;
        }

        if (!messageIterator.hasNext(DiscordUtils.USER_MENTION_PATTERN)) {
            return true;
        } else {
            // advance the scanner one step to get rid of user mention
            messageIterator.next();
        }

        User banUser = message.getMentionedUsers().get(0);
        Member banMember = guild.getMember(banUser);
        GuildController controller = guild.getController();

        if (!selfMember.hasPermission(Permission.BAN_MEMBERS)) {
            DiscordUtils.failMessage(bot, message, "I don't have enough permissions to do that!");
            return false;
        }

        if (user.getId().equals(banUser.getId())) {
            DiscordUtils.failMessage(bot, message, "You can't ban yourself, dummy!");
            return false;
        }

        if (!DiscordUtils.isBannable(banMember, selfMember)) {
            DiscordUtils.failMessage(bot, message, "I don't have enough permissions to do that!");
            return false;
        }

        Pair<String, Date> parsedReasonAndTime;

        try {
            parsedReasonAndTime = TextUtils.getTextAndTime(messageIterator);
        } catch (TextUtils.InvalidTimeInputException e) {
            DiscordUtils.failMessage(bot, message, "Invalid time argument. Please try again.");
            return false;
        } catch (TextUtils.TimeInputInPastException e) {
            DiscordUtils.failMessage(bot, message, "Your time argument was set for the past. Try again.\n" +
                                                               "If you're specifying a date, e.g. `30 December`, make sure you also write the year.");
            return false;
        }

        String text = parsedReasonAndTime.getLeft();
        String reason = text == null || text.equals("") ? "No reason specified" : text;
        Date expirationDate = parsedReasonAndTime.getRight();
        Date now = new Date();

        EmbedBuilder embed = new EmbedBuilder();
        embed.setTitle("Banned from " + guild.getName());
        embed.setColor(new Color(0x4286F4));
        embed.setDescription("You were banned from " + guild.getName());
        embed.addField("Reason:", reason, false);
        embed.addField("Banned until", expirationDate != null ? expirationDate.toString() : "Indefinitely", false);
        embed.setFooter("Banned by " + DiscordUtils.getUserTagAndId(user), null);
        embed.setTimestamp(now.toInstant());

        DiscordUtils.sendDM(banUser, embed.build());

        try {
            String auditLogReason = String.format("Banned by %s - %s", DiscordUtils.getUserTagAndId(user), reason);
            controller.ban(banMember, 0, auditLogReason).complete();
            DiscordUtils.successReact(bot, message);

            boolean expires = expirationDate != null;
            DSLContext database = bot.getDatabase();

            BanlistRecord record = database.insertInto(Tables.BANLIST,
                                                       Tables.BANLIST.USERID,
                                                       Tables.BANLIST.MODERATORUSERID,
                                                       Tables.BANLIST.GUILDID,
                                                       Tables.BANLIST.BANTIME,
                                                       Tables.BANLIST.EXPIRETIME,
                                                       Tables.BANLIST.REASON,
                                                       Tables.BANLIST.EXPIRES,
                                                       Tables.BANLIST.UNBANNED)
                                           .values(banUser.getId(),
                                                   user.getId(),
                                                   guild.getId(),
                                                   now.getTime() / 1000,
                                                   expirationDate != null ? expirationDate.getTime() / 1000 : 0,
                                                   reason,
                                                   expires,
                                                   false)
                                           .returning(Tables.BANLIST.ID)
                                           .fetchOne();

            int banId = record.getId();
            DiscordUtils.createModLogEntry(bot, shard, message, banMember, reason, "ban", banId, expirationDate, true);
            DiscordUtils.sendMessage(channel, "Banned " + DiscordUtils.getUserTagAndId(banUser));
        } catch (Exception e) {
            DiscordUtils.failMessage(bot, message, "Could not ban the specified user. Do I have enough permissions?");
        }

        return false;
    }
}

File: src/main/java/org/samoxive/safetyjim/discord/commands/Ban.kt
package org.samoxive.safetyjim.discord.commands

import java.awt.Color
import java.util.*
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.TextChannel
import net.dv8tion.jda.api.entities.User
import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent
import org.samoxive.safetyjim.database.BanEntity
import org.samoxive.safetyjim.database.BansTable
import org.samoxive.safetyjim.database.SettingsEntity
import org.samoxive.safetyjim.discord.*

suspend fun banAction(guild: Guild, channel: TextChannel?, settings: SettingsEntity, modUser: User, banUser: User, reason: String, expirationDate: Date?) {
    val now = Date()
    val embed = EmbedBuilder()
    embed.setTitle("Banned from ${guild.name}")
    embed.setColor(Color(0x4286F4))
    embed.setDescription("You were banned from ${guild.name}")
    embed.addField("Reason:", truncateForEmbed(reason), false)
    embed.addField("Banned until", expirationDate?.toString() ?: "Indefinitely", false)
    embed.setFooter("Banned by ${modUser.getUserTagAndId()}", null)
    embed.setTimestamp(now.toInstant())

    banUser.trySendMessage(embed.build())

    val auditLogReason = "Banned by ${modUser.getUserTagAndId()} - $reason"
    guild.ban(banUser, 0, auditLogReason).await()

    val expires = expirationDate != null

    BansTable.invalidatePreviousUserBans(guild, banUser)
    val record = BansTable.insertBan(
        BanEntity(
            userId = banUser.idLong,
            moderatorUserId = modUser.idLong,
            guildId = guild.idLong,
            banTime = now.time / 1000,
            expireTime = if (expirationDate != null) expirationDate.time / 1000 else 0,
            reason = reason,
            expires = expires,
            unbanned = false
        )
    )

    val banId = record.id
    createModLogEntry(guild, channel, settings, modUser, banUser, reason, ModLogAction.Ban, banId, expirationDate)
}

class Ban : Command() {
    override val usages = arrayOf("ban @user [reason] | [time] - bans the user with specific arguments. Both arguments can be omitted")

    override suspend fun run(bot: DiscordBot, event: GuildMessageReceivedEvent, settings: SettingsEntity, args: String): Boolean {
        val messageIterator = Scanner(args)

        val member = event.member!!
        val user = event.author
        val message = event.message
        val channel = event.channel
        val guild = event.guild
        val selfMember = guild.selfMember

        if (!member.hasPermission(Permission.BAN_MEMBERS)) {
            message.failMessage("You don't have enough permissions to execute this command! Required permission: Ban Members")
            return false
        }

        if (args.isEmpty()) {
            return true
        }

        val (searchResult, banUser) = messageIterator.findUser(message, true)
        if (searchResult == SearchUserResult.NOT_FOUND || (banUser == null)) {
            message.failMessage("Could not find the user to ban!")
            return false
        }

        if (searchResult == SearchUserResult.GUESSED) {
            message.askConfirmation(bot, banUser) ?: return false
        }

        val banMember = guild.getMember(banUser)

        if (!selfMember.hasPermission(Permission.BAN_MEMBERS)) {
            message.failMessage("I don't have enough permissions to do that!")
            return false
        }

        if (user == banUser) {
            message.failMessage("You can't ban yourself, dummy!")
            return false
        }

        if (banMember != null && !banMember.isBannableBy(selfMember)) {
            message.failMessage("I don't have enough permissions to do that!")
            return false
        }

        val parsedReasonAndTime = try {
            messageIterator.getTextAndTime()
        } catch (e: InvalidTimeInputException) {
            message.failMessage("Invalid time argument. Please try again.")
            return false
        } catch (e: TimeInputInPastException) {
            message.failMessage("Your time argument was set for the past. Try again.\nIf you're specifying a date, e.g. `30 December`, make sure you also write the year.")
            return false
        }

        val (text, expirationDate) = parsedReasonAndTime
        val reason = if (text == "") "No reason specified" else text

        try {
            banAction(guild, channel, settings, user, banUser, reason, expirationDate)
            message.successReact()
            channel.sendModActionConfirmationMessage(settings, "Banned ${banUser.getUserTagAndId()} ${getExpirationTextInChannel(expirationDate)}")
        } catch (e: Exception) {
            message.failMessage("Could not ban the specified user. Do I have enough permissions?")
        }

        return false
    }
}
